### DISCUSSION TITLE
Smooth Video Rendering

### DISCUSSION THREAD

**Discussion by superware** [AvaloniaUI/Avalonia#16836] [Q&A]
Hello,

I have a `GStreamer` pipeline ending with `d3d11videosink` drawing textures (video frames) into a `D3DImageEx` back buffer, this is super low CPU since no buffers are being copied. For reference, in `WPF`, the only way I manage to get smooth playback is by running a new `Dispatcher` in a different thread and using `HostVisual` and `VisualTarget` to connect the two visual trees ([details](https://dwayneneed.github.io/wpf/2007/04/26/multithreaded-ui-hostvisual.html)). Plus, no "airspace" issues - can draw elements on top of the video.

`GStreamer` is [very versatile and cross-platform](https://gstreamer.freedesktop.org/documentation/tutorials/basic/platform-specific-elements.html?gi-language=c), but it seems to be mainly up to `Avalonia` to allow rendering a video frame **immediately** (since the pipeline is syncing playback), anything else causes tearing, skewing and other artifacts. Is there a way to get smooth playback in `Avalonia`?

**Comment by maxkatz6** [comment]
Smooth video rendering comes to two possible implementations:
1. Use GPU interop, and render content directly into the texture (d3d, opengl or vulkan interop are supported by Avalonia, depending on platform). You should be able to use `glimagesink` and `d3d11videosink`. in Avalonia, see these samples https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop

2. On some other platforms you might need to use native video controls, in Avalonia it's possible via NativeControlHost - but airspace issue is there.

**Reply by superware** [reply]
Thanks @maxkatz6. IMHO a smooth video player/capabilities is a must for Avalonia, but it seems that currently most developers will simply define this as a limitation of the ecosystem.

I want to explore the GPU interop option (starting with Windows) since it has no airspace issues and it also feels more "right". Will really appreciate it if you can please reference me to a simpler example of what is minimally needed to allow `d3d11videosink` to draw on.

Thank you!

**Reply by superware** [reply]
So I managed to get video displayed ([gist](https://gist.github.com/superware/1ea60fa38aa05820a08a8b83274aabb6)), unfortunately still not smooth (a bit jumpy).

Is the code in `PostDraw` (being called on the GStreamer thread) the optimal way?

**Reply by maxkatz6** [reply]
`Dispatcher.UIThread.Invoke` is suboptimal there, yes. Since Avalonia render thread is separated from UI thread.

**Reply by superware** [reply]
My guess is WPF's `HostVisual` and `VisualTarget` are some kind of shortcut to the render thread or to the entire composed visual, there's no explicit thread marshaling and the result is totally smooth (see video below).

Comparing to Avalonia which is choppy in general and especially when interacting with other graphical controls such as [XAML Map Control](https://github.com/ClemensFischer/XAML-Map-Control) (used it on both).

https://github.com/user-attachments/assets/ea6d1441-da45-45a3-b118-194438bfd244

https://github.com/user-attachments/assets/5f8b7519-12fc-44fb-a2cc-5c1ae580287b

This is just a simple demo, I've been working on a full-blown Avalonia app for the past few months, I really prefer Avalonia for all the right reasons, but for me this must first be resolved. I hope it's possible.

**Reply by maxkatz6** [reply]
> Comparing to Avalonia which is choppy in general

It all comes down to implementation. Also remembering that WPF assemblies are pre-jited via ngen (at least for .net framework). For Avalonia you have a possibility to use Ready2Run but also even NativeAOT something that is not possible with WPF.

With our experience with XPF, we didn't experience chopines in Avalonia rendering engine, and old WPF apps sometimes even had a better performance, due to newer rendering stack (dx11 on windows, for example), or, again, NativeAOT.

**Reply by superware** [reply]
Let me rephrase, by "comparing to Avalonia which is choppy in general" I meant that the video plays a bit choppy even without interacting with that map control (although it's a bit better in x64 Release, I would even consider "production grade"), but panning the map causes a more noticeable effect. Please play the videos on desktop in full screen to appreciate the differences.

Can you please estimate why that interaction significantly affects video smoothness? Is there a way to somehow boost the video control render priority (or possibly lower other controls priority)? Obviously only the video is sensitive to exact rendering timing.

**Reply by maxkatz6** [reply]
> Can you please estimate why that interaction significantly affects video smoothness

Since video control uses dispatcher for each frame updates, it's very sensitive to any other UI operation. Pretty much my assumption, without running any code.

> Is there a way to somehow boost the video control render priority

Ideally - don't touch UI thread or dispatcher at all, when there is no need to. And keep everything on the render thread.
I can't really help with exact code, but Composition APIs and GPU interop is basically what you need. Something like CompositionCustomVisualHandler can also help, since you can trigger next render frame there directly.

**Reply by FoggyFinder** [reply]
> And keep everything on the render thread.

True. Yet I didn't find any sample to get access to render thread outside of Render method

**Reply by superware** [reply]
I looked into `CompositionCustomVisualHandler` but failed to fully understand it. There are actually only 3 inputs from `GStreamer`:

* The surface shared handler
* Resize event (width/height)
* "draw" event (surface updated)

I tried `UpdateWithKeyedMutexAsync` on the control's `Render` method (instead of on the UI thread) but then the method is only being called once.

Wish there was a simple example of the most performant way to render a DX surface. And since `GStreamer` is cross-platform with high composition, this might lead to a comprehensive solution for video playing/streaming in `Avalonia`.

**Reply by superware** [reply]
> > And keep everything on the render thread.
> 
> True. Yet I didn't find any sample to get access to render thread outside of Render method

@FoggyFinder Any progress?

**Reply by superware** [reply]
Hello @maxkatz6, I still can't manage to have a GStreamer video player to run smoothly, is there a new way? a new example for DX rendering? Maybe without going through the UI thread but directly to the Render thread? Thank you in advance!

**Reply by MikeCodesDotNET** [reply]
@superware you should probably just hold out for our upcoming media player that will be available to purchase very soon.

**Reply by FoggyFinder** [reply]
> @FoggyFinder Any progress?

@superware  no, I gave up. I'll probably wait when it will be fixed on Avalonia side rather than struggling to work it out

**Reply by superware** [reply]
> @superware you should probably just hold out for our upcoming media player that will be available to purchase very soon.

@MikeCodesDotNET Why would I want to purchase a media player if I've already implemented one which suites my needs? "A decade in the making, Avalonia is a mature and stable platform for building desktop, embedded, mobile, and web applications." sounds good, it doesn't say "you'll probably need to pay someone at some point" but rather another decade of maturity. Is the tide shifting? I'm asking for help with smoothly rendering frames, isn't this the right place/way?

**Reply by grokys** [reply]
If it meets your needs why are you asking for help?

**Reply by superware** [reply]
> If it meets your needs why are you asking for help?

Since I have a highly customized media pipeline, a generic media player most probably won't suffice. I've already invested time in its development and it works perfectly by its own, just trying since last August to get some DirectX interop/render thread help.

**Reply by toomasz** [reply]
@superware I'm also working on playing video streams from gstreamer in avalonia and exploring ways to integrate smooth video playback. Your approach with direct rendering from GStreamer to a D3D texture looks very promising.

I'd love to dig deeper — would you be willing to share the full code you used for this? I'm particularly interested in getting direct rendering from GStreamer to a D3D texture working as well.

Thanks in advance!

**Reply by superware** [reply]
@toomasz I tried contacting you via LinkedIn so we can discuss this further.

**Reply by Architrixs** [reply]
@superware @toomasz Hi guys, mind sharing some insights on how you guys achieved this, I have similar use case, and performance requirement. Any help is appreciated.

**Reply by superware** [reply]
@Architrixs Unfortunately I couldn't get good results - playback sync was very sensitive to what the UI thread was doing, so decided to skip Avalonia and stick with WPF ([pretty active](https://github.com/dotnet/wpf)). Using `HostVisual` I'm getting super smooth playback. Update me if you need help with that.
